src/charm.py

```diff
@@ -73,37 +94,112 @@ def _on_jenkins_pebble_ready(self, event: PebbleReadyEvent) -> None:
         """Configure and start Jenkins server.
 
         Args:
-            event: Event fired when pebble is ready.
+            event: The event fired when pebble is ready.
         """
         container = event.workload
         if not container or not container.can_connect():
             event.defer()
             return
 
-        self.unit.status = MaintenanceStatus("Configuring Jenkins.")
+        self.unit.status = MaintenanceStatus("Installing Jenkins.")
         # First Jenkins server start installs Jenkins server.
         container.add_layer(
-            "jenkins", self._get_pebble_layer(calculate_env(admin_configured=False)), combine=True
+            "jenkins",
+            self._get_pebble_layer(jenkins.calculate_env(admin_configured=False)),
+            combine=True,
         )
         container.replan()
         try:
-            wait_jenkins_ready()
-            unlock_jenkins(container)
+            jenkins.wait_ready()
+            self.unit.status = MaintenanceStatus("Configuring Jenkins.")
+            jenkins.bootstrap(
+                container,
+                self.state.jnlp_port,
+                self.state.num_executors,
+                self.state.plugins,
+            )
             # Second Jenkins server start restarts Jenkins to bypass Wizard setup.
             container.add_layer(
                 "jenkins",
-                self._get_pebble_layer(calculate_env(admin_configured=True)),
+                self._get_pebble_layer(jenkins.calculate_env(admin_configured=True)),
```

Comment:
Do you think `admin_configured` is state as well?

Reply:
Refactored to `is_bootstrapped` function in the `jenkins` module :), thanks for the suggestion.
As of having the container for dynamic state, I think we could pass in the `self.unit` for the `from_charm` function to be able to call `get_container(self.jenkins_service_name)` within. 
The reason for it is to avoid deadlocking of `charm.py`'s `_jenkins_container` property
```
 @property
    def _jenkins_container(self) -> Container:
        """The Jenkins workload container."""
        return self.unit.get_container(self.state.jenkins_service_name)
```
and `State` initialization
```
self.state = State(..., self._jenkins_container)
```
(note that the `_jenkins_container` depends on the `state.jenkins_service_name` and the state would depend on `_jenkins_container`).

Reply:
removed in favor of `container.restart`